mountpoint-s3-client/src/mock_client/throughput_client.rs

```diff
@@ -0,0 +1,214 @@
+        self.inner
+            .list_objects(bucket, continuation_token, delimiter, max_keys, prefix)
+            .await
+    }
+
+    async fn head_object(
+        &self,
+        bucket: &str,
+        key: &str,
+    ) -> ObjectClientResult<HeadObjectResult, HeadObjectError, Self::ClientError> {
+        self.inner.head_object(bucket, key).await
+    }
+
+    async fn put_object(
+        &self,
+        bucket: &str,
+        key: &str,
+        params: &PutObjectParams,
+    ) -> ObjectClientResult<Self::PutObjectRequest, PutObjectError, Self::ClientError> {
+        self.inner.put_object(bucket, key, params).await
+    }
+
+    async fn get_object_attributes(
+        &self,
+        bucket: &str,
+        key: &str,
+        max_parts: Option<usize>,
+        part_number_marker: Option<usize>,
+        object_attributes: &[ObjectAttribute],
+    ) -> ObjectClientResult<GetObjectAttributesResult, GetObjectAttributesError, Self::ClientError> {
+        self.inner
+            .get_object_attributes(bucket, key, max_parts, part_number_marker, object_attributes)
+            .await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Instant;
+
+    use futures::executor::block_on;
+    use futures::StreamExt;
+
+    use crate::mock_client::MockObject;
+
+    use super::*;
+
+    #[test]
+    fn ramp_throughput() {
+        const OBJECT_SIZE: usize = 128 * 1024 * 1024;
+        const ITERATIONS: usize = 1;
+
+        for rate_gbps in [0.5, 1.0, 2.0] {
+            for _ in 0..ITERATIONS {
+                let config = MockClientConfig {
+                    part_size: 8 * 1024 * 1024,
+                    bucket: "test_bucket".to_owned(),
+                    unordered_list_seed: None,
+                };
+                let client = ThroughputMockClient::new(config, rate_gbps);
+
+                client
+                    .inner
+                    .add_object("testfile", MockObject::ramp(0xaa, OBJECT_SIZE, ETag::for_tests()));
+
+                // Stream the entire object and drop it on the floor
+                let start = Instant::now();
+                let num_bytes = block_on(async move {
+                    let mut num_bytes = 0;
+                    let mut get = client.get_object("test_bucket", "testfile", None, None).await.unwrap();
+                    while let Some(part) = get.next().await {
+                        let (_offset, part) = part.unwrap();
+                        num_bytes += part.len();
+                    }
+                    num_bytes
+                });
+                let actual = start.elapsed().as_secs_f64();
+
+                assert_eq!(num_bytes, OBJECT_SIZE, "didn't stream entire object");
+
+                let expected = OBJECT_SIZE as f64 / (rate_gbps / 8.0 * 1.0e9);
+                assert!(
+                    actual > expected * 0.9,
+                    "too fast: rate_gbps={rate_gbps} actual={actual}s expected={expected}s"
+                );
```

Comment:
Should we also check that the client isn't _too_ slow?

Reply:
That's the check right below this one, but checking for "too slow" is pretty sensitive to slow hosts, so I'm leaving it commented out to avoid it becoming flaky in CI.